modules/oracledb/oracledb.go

```diff
@@ -0,0 +1,629 @@
+package oracledb
+
+import (
+	"database/sql"
+	"strconv"
+	"strings"
+
+	"github.com/netdata/go.d.plugin/modules"
+
+	"gopkg.in/goracle.v2"
+)
+
+const defaultDSN = "SYSTEM/Oracle12345@ORCL" // there is no such thing as default connstr/dsn, but...
+
+func init() {
+	creator := modules.Creator{
+		// TODO: open a discussion about API changes with @ilyam for that creator thing and more.
+		Create: func() modules.Module { return New(defaultDSN) },
+	}
+
+	modules.Register("oracledb", creator)
+}
+
+// OracleDB oracledb module.
+type OracleDB struct {
+	modules.Base
+	DSN string `json:"dsn,omitempty" yaml:"dsn,omitempty"`
+}
+
+// New creates OracleDB mod.
+func New(connString string) *OracleDB {
+	return &OracleDB{
+		DSN: defaultDSN,
+	}
+}
+
+// Cleanup performs cleanup.
+func (*OracleDB) Cleanup() {}
+
+// Init makes initialization of the OracleDB mod.
+func (m *OracleDB) Init() bool {
+	// test the connectivity here, this session will be not actually used.
+	// each metrics scraping will have its own session.
+	db, err := sql.Open("goracle", m.DSN)
+	if err != nil {
+		m.Errorf("error on opening a connection with the oracle database [%s]: %v", m.DSN, err)
+		return false
+	}
+
+	if err = db.Ping(); err != nil {
+		db.Close()
+		m.Errorf("error on pinging the oracle database [%s]: %v", m.DSN, err)
+		return false
+	}
+
+	if err = db.Close(); err != nil {
+		return false
+	}
+
+	// post Init debug info.
+	m.Debugf("using DSN [%s]", m.DSN)
+	return true
+}
+
+// Check makes check.
+func (m *OracleDB) Check() bool {
+	return len(m.Collect()) > 0
+}
+
+// Charts creates Charts.
+func (m *OracleDB) Charts() *Charts {
+	return charts.Copy()
+}
+
+// Collect collects health checks and metrics for OracleDB.
+func (m *OracleDB) Collect() map[string]int64 {
+	metrics := make(map[string]int64)
+
+	db, err := sql.Open("goracle", m.DSN)
```

Comment:
I see you reopen connection on every iteration. How about to reuse same connection and reopen it (reconnect) only in case of failures?

Reply:
Yes, opening a new connection on every itaration was not accidental, other metric systems do the same thing but if you want to re-use the connection I'll do it now, no problem.

Reply:
It's not something i want, if reusing same connection is more effective in terms of performance we should do it. If it is not open/close is ok too.

I think we don't need `ping` and it is ok to assume that connection is alive. If `db.query` returns an error (maybe some type of error, not an expert) the flow should be:

Reply:
✅